src/Microsoft.OData.Client/Metadata/ClientTypeUtil.cs

```diff
@@ -439,70 +442,80 @@ internal static PropertyInfo[] GetKeyPropertiesOnType(Type type)
         /// <returns>Returns the list of key properties defined on <paramref name="type"/>; null if <paramref name="type"/> is complex.</returns>
         internal static PropertyInfo[] GetKeyPropertiesOnType(Type type, out bool hasProperties)
         {
-            if (CommonUtil.IsUnsupportedType(type))
-            {
-                throw new InvalidOperationException(c.Strings.ClientType_UnsupportedType(type));
-            }
+            ODataTypeInfo typeInfo = GetODataTypeInfo(type);

```

Comment:
GetODataTypeInfo() should return null if type is new

Reply:
Get is intended to add to dict if a type is not present and then return it